### ISSUE TITLE
stgit cannot initialize branch under worktrees

### ISSUE THREAD

**Issue by jnsnow** [stacked-git/stgit#41]
I still need to investigate this, but I need to write it down before I forget.

When using worktrees, if you check out a branch that you have previously initialized with stgit, if you type `stg init`, everything will be "OK", if "OK" means a separate stack per-worktree.

If you type `stg series` instead, it will half-initialize a folder under `.git/worktrees/TREE/patches/BRANCH/` -- but then commands will fail, as it thinks it is both initialized and uninitialized simultaneously. `stg init` will fail, but `stg series` will assume we are uninitialized.

I think what we want is for `stg` to effectively *ignore* work trees and use the git-common-dir. git does not allow for one branch to be checked out simultaneously, so we don't need to worry about fighting ourselves for control over stg resources.

Note: `stg init` thinks we are properly initialized because of the presence of `.git/worktrees/TREE/patches/BRANCH`, but `stg series` fails because of the absence of the `applied` file.

**Comment by jnsnow** [comment]
Looking into this some more.

It's not clear to me if people are using stgit and worktrees in conjunction and *enjoy* the fact that a branch is decoupled from the patches that describe it. I think this is probably a bug. However, do I need to code in support for "recovering" the "old way" of doing things?

i.e. say you have a worktree and your patches are stored in the worktree git dir (project/.git/worktree/treename/patches/branchname) If I update stgit to use the COMMON git dir (project/.git/patches/branchname) instead, is this a meaningful breakage, or a bugfix?

It may be the case that we need to teach stgit to migrate files from worktrees over to the common dir, but what happens if there's a conflict there? this is potentially a mess.

**Comment by jnsnow** [comment]
One more thing:

The reason that (at present) the per-worktree support doesn't work correctly is because `update_to_current_format_version` only checks for the presence of the containing directory, but not necessarily the files within it (applied, unapplied, hidden, etc).

However, because the stgit format version data is stored in a per-branch way (using git-config data), it sees that we have previously initialized (in another worktree, somewhere else) and it decides to "upgrade" by creating the hidden patches file -- but none of the other files.

Shortly thereafter, it chokes because the files it expects to see are missing (no 'applied' or 'unapplied' files.)

So at the moment we have a mix of per-worktree and per-branch state information that stgit does not know how to interpret correctly in various edge-cases and situations.

per-worktree: Anything in the 'patches' folder.
per-branch: git-config metadata, such as version information.

**Comment by jpgrayson** [comment]
As of StGit 1.0, all of the stack metadata state is contained in the `<branch>.stgit` branch along with regular, auto-maintained references to patches of the form `refs/patches/<branch>/<patchname>`. So, no more `.git/patches` directory.

I did some initial testing with a worktree created by `git worktree add`. My observation is that StGit's behavior with a worktree seems to be reasonable now.

```sh
cd main-repo
git worktree add ../other
cd ../other
stg init
stg new -m p0
# edit something
stg refresh
...
```

The `stg init` initializes stack metadata on the `other.stgit` branch, as we want. And other StGit commands operate as expected on the `other` worktree's `other` and `other.stgit` branches. The state of the master branch and the main worktree are unaffected by StGit operations on the alternative worktree.

So I believe that this issue was fixed in StGit v1.0. This is a pleasant surprise for me; I was expecting that more would need to be done.

@jnsnow, I am wondering if you would be willing to give StGit 1.0 a try with worktrees? I'd like to know if you're seeing any inappropriate behaviors? Do you see any opportunities for StGit to better support worktrees?
